File: src/anomaly/graph_scorers.rs
use std::collections::HashMap;
use std::hash::Hash;

use crate::core::cluster::Cluster;
use crate::core::dataset::Dataset;
use crate::core::graph::Graph;
use crate::core::number::Number;
use crate::utils::helpers;

pub type ClusterScores<'a, T, U, D> = HashMap<&'a Cluster<'a, T, U, D>, f64>;
pub type InstanceScores = HashMap<usize, f64>;

pub trait GraphScorer<'a, T: Number, U: Number, D: Dataset<'a, T, U>>: Hash {
    fn call(&self, graph: &'a Graph<'a, T, U, D>) -> (ClusterScores<'a, T, U, D>, Vec<f64>) {
        let cluster_scores = {
            let mut cluster_scores = self.score_graph(graph);
            if self.normalize_on_clusters() {
                let (clusters, scores): (Vec<_>, Vec<_>) = cluster_scores.into_iter().unzip();
                cluster_scores = clusters
                    .into_iter()
                    .zip(helpers::normalize_1d(&scores).into_iter())
                    .collect();
            }
            cluster_scores
        };

        let instance_scores = {
            let mut instance_scores = self.inherit_scores(&cluster_scores);
            if !self.normalize_on_clusters() {
                let (indices, scores): (Vec<_>, Vec<_>) = instance_scores.into_iter().unzip();
                instance_scores = indices
                    .into_iter()
                    .zip(helpers::normalize_1d(&scores).into_iter())
                    .collect();
            }
            instance_scores
        };

        let scores_array = self.ordered_scores(&instance_scores);

        (cluster_scores, scores_array)
    }

    fn name(&self) -> &str;

    fn short_name(&self) -> &str;

    fn normalize_on_clusters(&self) -> bool;

    fn score_graph(&self, graph: &'a Graph<'a, T, U, D>) -> ClusterScores<'a, T, U, D>;

    fn inherit_scores(&self, scores: &ClusterScores<'a, T, U, D>) -> InstanceScores {
        scores
            .iter()
            .flat_map(|(&c, &s)| c.indices().iter().map(move |&i| (i, s)))
            .collect()
    }

    fn ordered_scores(&self, scores: &InstanceScores) -> Vec<f64> {
        let mut scores: Vec<_> = scores.iter().map(|(&i, &s)| (i, s)).collect();
        scores.sort_by_key(|(i, _)| *i);
        let (_, scores): (Vec<_>, Vec<f64>) = scores.into_iter().unzip();
        scores
    }
}

pub struct ClusterCardinality;

impl Hash for ClusterCardinality {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        "cluster_cardinality".hash(state)
    }
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> GraphScorer<'a, T, U, D> for ClusterCardinality {
    fn name(&self) -> &str {
        "cluster_cardinality"
    }

    fn short_name(&self) -> &str {
        "cc"
    }

    fn normalize_on_clusters(&self) -> bool {
        true
    }

    fn score_graph(&self, graph: &'a Graph<'a, T, U, D>) -> ClusterScores<'a, T, U, D> {
        graph
            .ordered_clusters()
            .iter()
            .map(|&c| (c, c.cardinality() as f64))
            .collect()
    }
}

pub struct ComponentCardinality;

impl Hash for ComponentCardinality {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        "component_cardinality".hash(state)
    }
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> GraphScorer<'a, T, U, D> for ComponentCardinality {
    fn name(&self) -> &str {
        "component_cardinality"
    }

    fn short_name(&self) -> &str {
        "sc"
    }

    fn normalize_on_clusters(&self) -> bool {
        true
    }

    fn score_graph(&self, graph: &'a Graph<'a, T, U, D>) -> ClusterScores<'a, T, U, D> {
        graph
            .find_component_clusters()
            .iter()
            .flat_map(|clusters| {
                let score = clusters.len() as f64;
                clusters.iter().map(move |&c| (c, score))
            })
            .collect()
    }
}

pub struct VertexDegree;

impl Hash for VertexDegree {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        "vertex_degree".hash(state)
    }
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> GraphScorer<'a, T, U, D> for VertexDegree {
    fn name(&self) -> &str {
        "vertex_degree"
    }

    fn short_name(&self) -> &str {
        "vd"
    }

    fn normalize_on_clusters(&self) -> bool {
        true
    }

    fn score_graph(&self, graph: &'a Graph<'a, T, U, D>) -> ClusterScores<'a, T, U, D> {
        graph
            .ordered_clusters()
            .iter()
            .map(|&c| (c, graph.unchecked_vertex_degree(c) as f64))
            .collect()
    }
}

pub struct ParentCardinality<'a, T: Number, U: Number, D: Dataset<'a, T, U>> {
    root: &'a Cluster<'a, T, U, D>,
    weight: Box<dyn (Fn(usize) -> f64) + Send + Sync>,
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> ParentCardinality<'a, T, U, D> {
    pub fn new(root: &'a Cluster<'a, T, U, D>) -> Self {
        let weight = Box::new(|d: usize| 1. / (d as f64).sqrt());
        Self { root, weight }
    }

    pub fn ancestry(&self, c: &'a Cluster<'a, T, U, D>) -> Vec<&'a Cluster<'a, T, U, D>> {
        c.history().into_iter().fold(vec![self.root], |mut ancestors, turn| {
            let last = ancestors.last().unwrap();
            let [left, right] = last.children().unwrap();
            let child = if turn { right } else { left };
            ancestors.push(child);
            ancestors
        })
    }
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> Hash for ParentCardinality<'a, T, U, D> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        "parent_cardinality".hash(state)
    }
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> GraphScorer<'a, T, U, D> for ParentCardinality<'a, T, U, D> {
    fn name(&self) -> &str {
        "parent_cardinality"
    }

    fn short_name(&self) -> &str {
        "pc"
    }

    fn normalize_on_clusters(&self) -> bool {
        true
    }

    fn score_graph(&self, graph: &'a Graph<'a, T, U, D>) -> ClusterScores<'a, T, U, D> {
        graph
            .ordered_clusters()
            .iter()
            .map(|&c| {
                let ancestry = self.ancestry(c);
                let score: f64 = ancestry
                    .iter()
                    .skip(1)
                    .zip(ancestry.iter())
                    .enumerate()
                    .map(|(i, (child, parent))| {
                        (self.weight)(i + 1) * parent.cardinality() as f64 / child.cardinality() as f64
                    })
                    .sum();
                (c, -score)
            })
            .collect()
    }
}

pub struct GraphNeighborhood {
    eccentricity_fraction: f64,
}

impl GraphNeighborhood {
    pub fn new(eccentricity_fraction: f64) -> Self {
        Self { eccentricity_fraction }
    }

    fn num_steps<'a, T: Number, U: Number, D: Dataset<'a, T, U>>(
        &self,
        graph: &'a Graph<'a, T, U, D>,
        c: &'a Cluster<'a, T, U, D>,
    ) -> usize {
        let steps = graph.unchecked_eccentricity(c) as f64 * self.eccentricity_fraction;
        1 + steps as usize
    }
}

impl Hash for GraphNeighborhood {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        "graph_neighborhood".hash(state)
    }
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> GraphScorer<'a, T, U, D> for GraphNeighborhood {
    fn name(&self) -> &str {
        "graph_neighborhood"
    }

    fn short_name(&self) -> &str {
        "gn"
    }

    fn normalize_on_clusters(&self) -> bool {
        true
    }

    fn score_graph(&self, graph: &'a Graph<'a, T, U, D>) -> ClusterScores<'a, T, U, D> {
        graph
            .ordered_clusters()
            .iter()
            .map(|&c| {
                let steps = self.num_steps(graph, c);
                // TODO: Do we need +1?
                let score = (0..steps + 1)
                    .zip(graph.unchecked_frontier_sizes(c).iter())
                    .fold(0, |score, (_, &size)| score + size);
                (c, -(score as f64))
            })
            .collect()
    }
}

#[allow(dead_code)]
pub struct StationaryProbabilities {
    num_steps: usize,
}

impl Hash for StationaryProbabilities {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        "stationary_probabilities".hash(state)
    }
}

impl StationaryProbabilities {
    pub fn new(num_steps: usize) -> Self {
        Self { num_steps }
    }
}

impl<'a, T: Number, U: Number, D: Dataset<'a, T, U>> GraphScorer<'a, T, U, D> for StationaryProbabilities {
    fn name(&self) -> &str {
        "stationary_probabilities"
    }

    fn short_name(&self) -> &str {
        "sp"
    }

    fn normalize_on_clusters(&self) -> bool {
        true
    }

    #[allow(unused_variables)]
    fn score_graph(&self, graph: &'a Graph<'a, T, U, D>) -> ClusterScores<'a, T, U, D> {
        todo!()
    }
}

File: python/abd_clam/anomaly_detection/graph_scorers.py
"""Graph anomaly detection algorithms."""

import abc
import typing

import numpy

from .. import core
from ..utils import constants
from ..utils import helpers

logger = helpers.make_logger(__name__)

ClusterScores = dict[core.Cluster, float]
InstanceScores = dict[int, float]

# TODO: Look at DSD (diffusion-state-distance) as a possible individual-method.
# TODO: Specialize `should_be_fast` for each algorithm based on complexity analysis
# and graph properties.


class GraphScorer(abc.ABC):
    """A base class for all anomaly detection algorithms."""

    def __call__(self, g: core.Graph) -> tuple[ClusterScores, numpy.ndarray]:
        """Use this scorer to generate normalized anomaly scores."""
        logger.info(
            f"Running method {self.name} on a "
            "graph with {g.vertex_cardinality} clusters.",
        )

        cluster_scores = self.score_graph(g)
        if self.normalize_on_clusters:
            cluster_scores = self.__normalize_cluster_scores(cluster_scores)

        instance_scores = self.inherit_scores(cluster_scores)
        if not self.normalize_on_clusters:
            instance_scores = self.__normalize_instance_scores(instance_scores)

        scores_array = self.ordered_scores(g, instance_scores)
        return cluster_scores, scores_array

    @abc.abstractmethod
    def __hash__(self) -> int:
        """A way to uniquely identify each `GraphScorer` object."""
        pass

    @property
    @abc.abstractmethod
    def name(self) -> str:
        """Full name of the algorithm."""
        pass

    @property
    @abc.abstractmethod
    def short_name(self) -> str:
        """Abbreviated name of the algorithm."""
        pass

    @property
    @abc.abstractmethod
    def normalize_on_clusters(self) -> bool:
        """Whether to normalize scores for clusters or for instances."""
        pass

    @property
    @abc.abstractmethod
    def normalization_mode(self) -> helpers.NormalizationMode:
        """What normalization method to use."""
        pass

    @abc.abstractmethod
    def score_graph(self, g: core.Graph) -> ClusterScores:
        """The main method of the algorithm.

        This should return a dictionary whose keys are `Clusters` in the `Graph`
        and whose values are the pre-normalization anomaly scores.
        """
        pass

    @abc.abstractmethod
    def should_be_fast(self, g: core.Graph) -> bool:
        """Whether this algorithm is expected to run in a reasonably short time.

        This estimate method should made quickly.
        """
        pass

    def __normalize_cluster_scores(
        self,
        scores: ClusterScores,
    ) -> ClusterScores:
        """Normalize scores to be in the [0, 1] range."""
        [keys, scores] = list(zip(*scores.items()))  # type: ignore[assignment]
        scores = helpers.normalize(
            numpy.asarray(scores, dtype=numpy.float32),
            mode=self.normalization_mode,
        )
        return dict(zip(keys, (float(x) for x in scores)))  # type: ignore[arg-type]

    def __normalize_instance_scores(
        self,
        scores: InstanceScores,
    ) -> InstanceScores:
        """Normalize scores to be in the [0, 1] range."""
        [keys, scores] = list(zip(*scores.items()))  # type: ignore[assignment]
        scores = helpers.normalize(
            numpy.asarray(scores, dtype=numpy.float32),
            mode=self.normalization_mode,
        )
        return dict(zip(keys, (float(x) for x in scores)))  # type: ignore[arg-type]

    @staticmethod
    def inherit_scores(scores: ClusterScores) -> InstanceScores:
        """Inherit cluster scores to instance scores."""
        return {i: s for c, s in scores.items() for i in c.indices}

    @staticmethod
    def ordered_scores(g: core.Graph, scores: InstanceScores) -> numpy.ndarray:
        """Return scores in the order of the graph's vertices."""
        [indices, scores_] = list(zip(*sorted(scores.items())))
        if set(indices) != {i for c in g.clusters for i in c.indices}:
            msg = "Graph invariant violated."
            raise ValueError(msg)
        return numpy.asarray(scores_, dtype=numpy.float32)


class ClusterCardinality(GraphScorer):
    """A method that scores clusters by their cardinality.

    The smaller the cluster, the more anomalous it is.
    """

    def __hash__(self) -> int:
        """A way to uniquely identify each `GraphScorer` object by its name."""
        return hash(self.name)

    @property
    def name(self) -> str:
        """Full name of the algorithm."""
        return "cluster_cardinality"

    @property
    def short_name(self) -> str:
        """Abbreviated name of the algorithm."""
        return "cc"

    @property
    def normalize_on_clusters(self) -> bool:
        """Whether to normalize scores for clusters."""
        return True

    @property
    def normalization_mode(self) -> helpers.NormalizationMode:
        """The normalization mode to use."""
        return "gaussian"

    def should_be_fast(self, _: core.Graph) -> bool:
        """This method is expected to be fast for all graphs."""
        return True

    def score_graph(self, g: core.Graph) -> ClusterScores:
        """The main method of the algorithm."""
        return {c: -c.cardinality for c in g.clusters}


class ComponentCardinality(GraphScorer):
    """A method that scores clusters by their component's cardinality.

    The smaller the component, the more anomalous its clusters are.
    """

    def __hash__(self) -> int:
        """A way to uniquely identify each `GraphScorer` object by its name."""
        return hash(self.name)

    @property
    def name(self) -> str:
        """Full name of the algorithm."""
        return "component_cardinality"

    @property
    def short_name(self) -> str:
        """Abbreviated name of the algorithm."""
        return "sc"

    @property
    def normalize_on_clusters(self) -> bool:
        """Whether to normalize scores for clusters."""
        return True

    @property
    def normalization_mode(self) -> helpers.NormalizationMode:
        """The normalization mode to use."""
        return "gaussian"

    def should_be_fast(self, _: core.Graph) -> bool:
        """This method is expected to be fast for all graphs."""
        return True

    def score_graph(self, g: core.Graph) -> ClusterScores:
        """The main method of the algorithm."""
        return {
            c: -component.vertex_cardinality
            for component in g.components
            for c in component.clusters
        }


class VertexDegree(GraphScorer):
    """A method that scores clusters by their vertex degree.

    The smaller the degree, the more anomalous the cluster is.
    """

    def __hash__(self) -> int:
        """A way to uniquely identify each `GraphScorer` object by its name."""
        return hash(self.name)

    @property
    def name(self) -> str:
        """Full name of the algorithm."""
        return "vertex_degree"

    @property
    def short_name(self) -> str:
        """Abbreviated name of the algorithm."""
        return "vd"

    @property
    def normalize_on_clusters(self) -> bool:
        """Whether to normalize scores for clusters."""
        return True

    @property
    def normalization_mode(self) -> helpers.NormalizationMode:
        """The normalization mode to use."""
        return "gaussian"

    def should_be_fast(self, _: core.Graph) -> bool:
        """This method is expected to be fast for all graphs."""
        return True

    def score_graph(self, g: core.Graph) -> ClusterScores:
        """The main method of the algorithm."""
        return {c: -g.vertex_degree(c) for c in g.clusters}


class ParentCardinality(GraphScorer):
    """A method that scores clusters by their parent's cardinality.

    The smaller the cluster relative to its parent, the more anomalous the cluster is.
    """

    def __init__(self, depth_weight: typing.Callable[[int], float]) -> None:
        """Initialize the scorer."""
        self.weight = depth_weight

    def __hash__(self) -> int:
        """A way to uniquely identify each `GraphScorer` object by its name."""
        return hash(self.name)

    @property
    def name(self) -> str:
        """Full name of the algorithm."""
        return "parent_cardinality"

    @property
    def short_name(self) -> str:
        """Abbreviated name of the algorithm."""
        return "pc"

    @property
    def normalize_on_clusters(self) -> bool:
        """Whether to normalize scores for clusters."""
        return True

    @property
    def normalization_mode(self) -> helpers.NormalizationMode:
        """The normalization mode to use."""
        return "gaussian"

    def should_be_fast(self, _: core.Graph) -> bool:
        """This method is expected to be fast for all graphs."""
        return True

    def score_graph(self, g: core.Graph) -> ClusterScores:
        """The main method of the algorithm."""
        scores = {c: 0.0 for c in g.clusters}

        for c in g.clusters:
            ancestry = c.ancestry
            for i in range(1, len(ancestry)):
                scores[c] += (
                    self.weight(i)
                    * ancestry[i - 1].cardinality
                    / ancestry[i].cardinality
                )

        return {c: scores[c] for c in g.clusters}


class GraphNeighborhood(GraphScorer):
    """A method that scores clusters by the number of clusters in their neighborhood.

    The smaller the neighborhood, the more anomalous the cluster is.
    """

    def __init__(self, eccentricity_fraction: float) -> None:
        """Initialize the scorer."""
        if not (0 < eccentricity_fraction <= 1):
            msg = (
                "eccentricity fraction must be in the (0, 1] range. "
                "Got {eccentricity_fraction:.2e} instead."
            )
            raise ValueError(
                msg,
            )

        self.eccentricity_fraction = eccentricity_fraction

    def __hash__(self) -> int:
        """A way to uniquely identify each `GraphScorer` object by its name."""
        return hash(self.name)

    @property
    def name(self) -> str:
        """Full name of the algorithm."""
        return "graph_neighborhood"

    @property
    def short_name(self) -> str:
        """Abbreviated name of the algorithm."""
        return "gn"

    @property
    def normalize_on_clusters(self) -> bool:
        """Whether to normalize scores for clusters."""
        return True

    @property
    def normalization_mode(self) -> helpers.NormalizationMode:
        """The normalization mode to use."""
        return "gaussian"

    def should_be_fast(self, g: core.Graph) -> bool:
        """This method is expected to be fast for small graphs."""
        return g.vertex_cardinality < 256

    def num_steps(self, g: core.Graph, c: core.Cluster) -> int:
        """The number of steps to take away from the cluster."""
        return int(g.eccentricity(c) * self.eccentricity_fraction) + 1

    def score_graph(self, g: core.Graph) -> ClusterScores:
        """The main method of the algorithm."""
        step_sizes: dict[core.Cluster, list[int]] = {
            c: g.frontier_sizes(c)[: self.num_steps(g, c)] for c in g.clusters
        }
        return {c: -float(sum(sizes)) for c, sizes in step_sizes.items()}


class StationaryProbabilities(GraphScorer):
    """A method that scores clusters by their stationary probabilities.

    The stationary probability of a cluster is the probability that a random walk
    starting from a vertex in the cluster will end up in the cluster after a large
    number of steps.

    The smaller the stationary probability, the more anomalous the cluster is.
    """

    def __init__(self, steps: int) -> None:
        """Initialize the scorer."""
        self.steps = steps

    def __hash__(self) -> int:
        """A way to uniquely identify each `GraphScorer` object by its name."""
        return hash(self.name)

    @property
    def name(self) -> str:
        """Full name of the algorithm."""
        return "stationary_probabilities"

    @property
    def short_name(self) -> str:
        """Abbreviated name of the algorithm."""
        return "sp"

    @property
    def normalize_on_clusters(self) -> bool:
        """Whether to normalize scores for clusters."""
        return True

    @property
    def normalization_mode(self) -> helpers.NormalizationMode:
        """The normalization mode to use."""
        return "gaussian"

    def should_be_fast(self, _: core.Graph) -> bool:
        """This method is expected to be fast for small graphs."""
        return True

    def score_graph(self, g: core.Graph) -> ClusterScores:
        """The main method of the algorithm."""
        scores = {}

        for component in g.components:
            if component.vertex_cardinality > 1:
                clusters, matrix = component.distance_matrix

                # TODO: Figure out how to go until convergence.
                sums = numpy.sum(matrix, axis=1)[:, None] + constants.EPSILON
                matrix = matrix / sums
                for _ in range(self.steps):
                    matrix = numpy.linalg.matrix_power(matrix, 2)

                scores.update(
                    dict(zip(clusters, numpy.sum(matrix, axis=0))),
                )

            else:  # Component with only one Cluster means the Cluster is anomalous.
                scores.update({c: 0 for c in component.clusters})

        scores = {c: -s for c, s in scores.items()}

        return scores


__all__ = [
    "ClusterScores",
    "InstanceScores",
    "GraphScorer",
    "ClusterCardinality",
    "ComponentCardinality",
    "VertexDegree",
    "ParentCardinality",
    "GraphNeighborhood",
    "StationaryProbabilities",
]
